Sources/BSON/BSONDecoder.swift

```diff
@@ -165,6 +154,33 @@ public class BSONDecoder {
         try self.decode(type, from: BSONDocument(fromBSON: data))
     }
 
+    /**
+     * Decodes a top-level value of the given type from the given BSON.
+     *
+     * - Parameter type: The type of the value to decode.
+     * - Parameter document: The BSON to decode from.
+     * - Returns: A value of the requested type.
+     * - Throws: `DecodingError` if any value throws an error during decoding.
+     */
+    internal func decode<T: Decodable>(_ type: T.Type, fromBSON bson: BSON) throws -> T {
+        // if the requested type is `BSON` we're done
+        if let b = bson as? T {
+            return b
+        }
+        let _decoder = _BSONDecoder(referencing: bson, options: self.options)
+        do {
+            return try type.init(from: _decoder)
+        } catch let error as BSONErrorProtocol {
+            let unknownErrorMessage = "Unknown Error occurred while decoding BSON"
+            throw DecodingError.dataCorrupted(
+                DecodingError.Context(
+                    codingPath: [],
```

Comment:
I know this is a copy paste, but now that I am looking at it I think it would make sense to use `_decoder.codingPath` here. if we encounter a BSON error it did not necessarily occur at the top level.

Reply:
ok